### PR TITLE
make first part of user agent header configurable

### PR DESCRIPTION
**Pull request by rino-kadijk**
This PR adds the configuration option to prefix the User Agent header with a custom appId. This allows the server to see where the queries originate from.

https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/User-Agent#library_and_net_tool_ua_strings

### CODE REVIEW

**Review by gortiz — APPROVED** [review]
LGTM

File: pinot-clients/pinot-java-client/src/main/java/org/apache/pinot/client/JsonAsyncHttpPinotClientTransport.java [hunk 1]
```diff
@@ -66,7 +66,8 @@ public JsonAsyncHttpPinotClientTransport() {
   }
 
   public JsonAsyncHttpPinotClientTransport(Map<String, String> headers, String scheme,
-    @Nullable SSLContext sslContext, ConnectionTimeouts connectionTimeouts, TlsProtocols tlsProtocols) {
+                                           @Nullable SSLContext sslContext, ConnectionTimeouts connectionTimeouts,
```
Comment (Jackie-Jiang) [inline pinot-clients/pinot-java-client/src/main/java/org/apache/pinot/client/JsonAsyncHttpPinotClientTransport.java:69]: (format) Please follow the [Pinot Style](https://docs.pinot.apache.org/developers/developers-and-contributors/code-setup#setup-ide). Same for other changes

Comment (Jackie-Jiang) [reply pinot-clients/pinot-java-client/src/main/java/org/apache/pinot/client/JsonAsyncHttpPinotClientTransport.java:69]: Annotate `appId` as `@Nullable`

File: pinot-clients/pinot-java-client/src/main/java/org/apache/pinot/client/JsonAsyncHttpPinotClientTransportFactory.java [hunk 2]
```diff
@@ -43,13 +43,15 @@ public class JsonAsyncHttpPinotClientTransportFactory implements PinotClientTran
   private int _readTimeoutMs = Integer.parseInt(DEFAULT_BROKER_READ_TIMEOUT_MS);
   private int _connectTimeoutMs = Integer.parseInt(DEFAULT_BROKER_READ_TIMEOUT_MS);
   private int _handshakeTimeoutMs = Integer.parseInt(DEFAULT_BROKER_HANDSHAKE_TIMEOUT_MS);
+  private String _appId = "";
```
Comment (Jackie-Jiang) [inline pinot-clients/pinot-java-client/src/main/java/org/apache/pinot/client/JsonAsyncHttpPinotClientTransportFactory.java:46]: Suggest setting it to `null` if not available

File: pinot-clients/pinot-java-client/src/main/java/org/apache/pinot/client/JsonAsyncHttpPinotClientTransportFactory.java [hunk 3]
```diff
@@ -94,6 +96,7 @@ public JsonAsyncHttpPinotClientTransportFactory withConnectionProperties(Propert
             DEFAULT_BROKER_CONNECT_TIMEOUT_MS));
     _handshakeTimeoutMs = Integer.parseInt(properties.getProperty("brokerHandshakeTimeoutMs",
             DEFAULT_BROKER_HANDSHAKE_TIMEOUT_MS));
+    _appId = properties.getProperty("appId", "");
```
Comment (Jackie-Jiang) [inline pinot-clients/pinot-java-client/src/main/java/org/apache/pinot/client/JsonAsyncHttpPinotClientTransportFactory.java:99]: Set it to `null` if not available

File: pinot-clients/pinot-java-client/src/main/java/org/apache/pinot/client/utils/ConnectionUtils.java [hunk 4]
```diff
@@ -56,14 +56,18 @@ public static SSLContext getSSLContextFromProperties(Properties properties) {
   }
 
 
-  public static String getUserAgentVersionFromClassPath(String userAgentKey) {
+  public static String getUserAgentVersionFromClassPath(String userAgentKey, String appId) {
```
Comment (Jackie-Jiang) [inline pinot-clients/pinot-java-client/src/main/java/org/apache/pinot/client/utils/ConnectionUtils.java:59]:
```suggestion
  public static String getUserAgentVersionFromClassPath(String userAgentKey, @Nullable String appId) {
```

File: pinot-clients/pinot-java-client/src/main/java/org/apache/pinot/client/utils/ConnectionUtils.java [hunk 5]
```diff
@@ -56,14 +56,18 @@ public static SSLContext getSSLContextFromProperties(Properties properties) {
...
     Properties userAgentProperties = new Properties();
     try {
       userAgentProperties.load(ConnectionUtils.class.getClassLoader()
           .getResourceAsStream("version.properties"));
     } catch (IOException e) {
       LOGGER.warn("Unable to set user agent version");
     }
-    return userAgentProperties.getProperty(userAgentKey, "pinot-java");
+    String userAgentFromProperties = userAgentProperties.getProperty(userAgentKey, "unknown");
+    if (appId != null && appId.length() > 0) {
```
Comment (Jackie-Jiang) [inline pinot-clients/pinot-java-client/src/main/java/org/apache/pinot/client/utils/ConnectionUtils.java:68]: (nit) StringUtils.isNotEmpty(appId)

File: pinot-clients/pinot-jdbc-client/src/main/java/org/apache/pinot/client/controller/PinotControllerTransport.java [hunk 6]
```diff
@@ -52,7 +52,8 @@ public class PinotControllerTransport {
 
 
   public PinotControllerTransport(Map<String, String> headers, String scheme,
-    @Nullable SSLContext sslContext, ConnectionTimeouts connectionTimeouts, TlsProtocols tlsProtocols) {
+                                  @Nullable SSLContext sslContext, ConnectionTimeouts connectionTimeouts,
```
Comment (Jackie-Jiang) [inline pinot-clients/pinot-jdbc-client/src/main/java/org/apache/pinot/client/controller/PinotControllerTransport.java:55]: Same comments as in `JsonAsyncHttpPinotClientTransport`

File: pinot-clients/pinot-jdbc-client/src/main/java/org/apache/pinot/client/controller/PinotControllerTransportFactory.java [hunk 7]
```diff
@@ -41,12 +41,13 @@ public class PinotControllerTransportFactory {
   private int _readTimeoutMs = Integer.parseInt(DEFAULT_CONTROLLER_READ_TIMEOUT_MS);
   private int _connectTimeoutMs = Integer.parseInt(DEFAULT_CONTROLLER_CONNECT_TIMEOUT_MS);
   private int _handshakeTimeoutMs = Integer.parseInt(DEFAULT_CONTROLLER_HANDSHAKE_TIMEOUT_MS);
+  private String _appId = "";
```
Comment (Jackie-Jiang) [inline pinot-clients/pinot-jdbc-client/src/main/java/org/apache/pinot/client/controller/PinotControllerTransportFactory.java:44]: Same comments as in `JsonAsyncHttpPinotClientTransportFactory`

**Review by Jackie-Jiang — COMMENTED** [review]
LGTM in general. Minor comments

**Comment by rino-kadijk** [comment]
> LGTM in general. Minor comments

will fix, should I trim the appId to 32 characters?

**Comment by Jackie-Jiang** [comment]
Is there a constraint on the appId length? If so, we should fail it when user configure it with a too long value

File: pinot-clients/pinot-java-client/src/main/java/org/apache/pinot/client/BrokerCache.java [hunk 8]
```diff
@@ -104,6 +104,7 @@ public BrokerCache(Properties properties, String controllerUrl) {
         DEFAULT_CONTROLLER_CONNECT_TIMEOUT_MS));
     int handshakeTimeoutMs = Integer.parseInt(properties.getProperty("controllerHandshakeTimeoutMs",
         DEFAULT_CONTROLLER_HANDSHAKE_TIMEOUT_MS));
+    String appId = properties.getProperty("appId", DEFAULT_CONTROLLER_HANDSHAKE_TIMEOUT_MS);
```
Comment (Jackie-Jiang) [inline pinot-clients/pinot-java-client/src/main/java/org/apache/pinot/client/BrokerCache.java:107]:
```suggestion
    String appId = properties.getProperty("appId");
```

File: pinot-clients/pinot-java-client/src/main/java/org/apache/pinot/client/JsonAsyncHttpPinotClientTransport.java (hunk 1 above)
Comment (Jackie-Jiang) [reply pinot-clients/pinot-java-client/src/main/java/org/apache/pinot/client/JsonAsyncHttpPinotClientTransport.java:73]: The format is still not correct. Please update the IDE style setting and reformat

File: pinot-clients/pinot-java-client/src/main/java/org/apache/pinot/client/JsonAsyncHttpPinotClientTransportFactory.java (hunk 3 above)
Comment (Jackie-Jiang) [reply pinot-clients/pinot-java-client/src/main/java/org/apache/pinot/client/JsonAsyncHttpPinotClientTransportFactory.java:101]: ^^

File: pinot-clients/pinot-jdbc-client/src/main/java/org/apache/pinot/client/controller/PinotControllerTransport.java [hunk 9]
```diff
@@ -64,21 +66,26 @@ public PinotControllerTransport(Map<String, String> headers, String scheme,
...
     try {
       userAgentProperties.load(PinotControllerTransport.class.getClassLoader()
               .getResourceAsStream("version.properties"));
     } catch (IOException e) {
       LOGGER.warn("Unable to set user agent version");
     }
-    return userAgentProperties.getProperty("ua", "unknown");
+    String userAgentFromProperties = userAgentProperties.getProperty("ua", "unknown");
+    if (StringUtils.isNotEmpty(appId)) {
+      return appId.substring(0, Math.min(org.apache.pinot.client.utils.ConnectionUtils.APP_ID_MAX_CHARS,
```
Comment (Jackie-Jiang) [inline pinot-clients/pinot-jdbc-client/src/main/java/org/apache/pinot/client/controller/PinotControllerTransport.java:85]: (minor) import `ConnectionUtils`

File: pinot-clients/pinot-jdbc-client/src/main/java/org/apache/pinot/client/controller/PinotControllerTransportFactory.java [hunk 10]
```diff
@@ -80,6 +81,7 @@ public PinotControllerTransportFactory withConnectionProperties(Properties prope
             DEFAULT_CONTROLLER_CONNECT_TIMEOUT_MS));
     _handshakeTimeoutMs = Integer.parseInt(properties.getProperty("controllerHandshakeTimeoutMs",
             DEFAULT_CONTROLLER_HANDSHAKE_TIMEOUT_MS));
+    _appId = properties.getProperty("appId", "");
```
Comment (Jackie-Jiang) [inline pinot-clients/pinot-jdbc-client/src/main/java/org/apache/pinot/client/controller/PinotControllerTransportFactory.java:84]: Same here, we want to set `null` when it is not configured
```suggestion
    _appId = properties.getProperty("appId");
```

File: pinot-clients/pinot-java-client/src/main/java/org/apache/pinot/client/JsonAsyncHttpPinotClientTransport.java [hunk 11]
```diff
@@ -70,7 +70,8 @@ public JsonAsyncHttpPinotClientTransport() {
   }
 
   public JsonAsyncHttpPinotClientTransport(Map<String, String> headers, String scheme, String extraOptionString,
-    @Nullable SSLContext sslContext, ConnectionTimeouts connectionTimeouts, TlsProtocols tlsProtocols) {
+                                           @Nullable SSLContext sslContext, ConnectionTimeouts connectionTimeouts,
```
Comment (Jackie-Jiang) [inline pinot-clients/pinot-java-client/src/main/java/org/apache/pinot/client/JsonAsyncHttpPinotClientTransport.java:73]: Please reformat this file

**Review by Jackie-Jiang — COMMENTED** [review]
Please take a look at the failed tests

**Comment by rino-kadijk** [comment]
> Please take a look at the failed tests

I don't see anything related to this failure in this PR:

```java.lang.NullPointerException
[20596](https://github.com/apache/pinot/actions/runs/3266516098/jobs/5370354908#step:5:20597)
	at org.apache.pinot.segment.local.io.readerwriter.RealtimeIndexOffHeapMemoryManager.close(RealtimeIndexOffHeapMemoryManager.java:109)
[20597](https://github.com/apache/pinot/actions/runs/3266516098/jobs/5370354908#step:5:20598)
	at org.apache.pinot.segment.local.indexsegment.mutable.MutableSegmentImpl.destroy(MutableSegmentImpl.java:1187)
[20598](https://github.com/apache/pinot/actions/runs/3266516098/jobs/5370354908#step:5:20599)
	at org.apache.pinot.core.data.manager.realtime.LLRealtimeSegmentDataManager.destroy(LLRealtimeSegmentDataManager.java:1213)
[20599](https://github.com/apache/pinot/actions/runs/3266516098/jobs/5370354908#step:5:20600)
	at org.apache.pinot.core.data.manager.realtime.RealtimeTableDataManager.doShutdown(RealtimeTableDataManager.java:210)
[20600](https://github.com/apache/pinot/actions/runs/3266516098/jobs/5370354908#step:5:20601)
	at org.apache.pinot.core.data.manager.BaseTableDataManager.shutDown(BaseTableDataManager.java:190)
[20601](https://github.com/apache/pinot/actions/runs/3266516098/jobs/5370354908#step:5:20602)
	at org.apache.pinot.server.starter.helix.HelixInstanceDataManager.shutDown(HelixInstanceDataManager.java:155)
[20602](https://github.com/apache/pinot/actions/runs/3266516098/jobs/5370354908#step:5:20603)
	at org.apache.pinot.server.starter.ServerInstance.shutDown(ServerInstance.java:261)
[20603](https://github.com/apache/pinot/actions/runs/3266516098/jobs/5370354908#step:5:20604)
	at org.apache.pinot.server.starter.helix.BaseServerStarter.stop(BaseServerStarter.java:624)
[20604](https://github.com/apache/pinot/actions/runs/3266516098/jobs/5370354908#step:5:20605)
	at org.apache.pinot.integration.tests.ClusterTest.stopServer(ClusterTest.java:276)
[20605](https://github.com/apache/pinot/actions/runs/3266516098/jobs/5370354908#step:5:20606)
	at org.apache.pinot.integration.tests.UrlAuthRealtimeIntegrationTest.tearDown(UrlAuthRealtimeIntegrationTest.java:91)```

**Review by Jackie-Jiang — APPROVED** [review]

**Pull request merged**